lib/java/desktop-environment/src/main/java/org/enso/desktopenvironment/LinuxTrash.java

```diff
@@ -0,0 +1,220 @@
+
+        try {
+          Files.move(
+              path,
+              trashDir.resolve(PATH_FILES).resolve(trashInfo.fileName),
+              StandardCopyOption.ATOMIC_MOVE);
+          return true;
+        } catch (IOException e) {
+          boolean isSuccessful;
+          if (Files.isDirectory(path)) {
+            isSuccessful =
+                moveDirectoryToDirectory(path, trashDir.resolve(PATH_FILES), trashInfo.fileName);
+          } else {
+            isSuccessful =
+                moveFileToDirectory(path, trashDir.resolve(PATH_FILES), trashInfo.fileName);
+          }
+
+          if (!isSuccessful) {
+            FileUtils.deleteQuietly(trashInfo.path.toFile());
+          }
+
+          return isSuccessful;
+        }
+      } catch (IOException e) {
+        return false;
+      }
+
+    } else {
+      return false;
+    }
+  }
+
+  private static boolean moveFileToDirectory(Path from, Path to, String fileName) {
+    var source = from.toFile();
+    var destination = to.resolve(fileName).toFile();
+
+    try {
+      FileUtils.copyFile(source, destination);
+      FileUtils.delete(source);
+
+      return true;
+    } catch (IOException e) {
+      FileUtils.deleteQuietly(destination);
+      return false;
+    }
+  }
+
+  private static boolean moveDirectoryToDirectory(Path from, Path to, String fileName) {
+    var source = from.toFile();
+    var destination = to.resolve(fileName).toFile();
+    try {
+      FileUtils.copyDirectory(source, destination);
+      FileUtils.deleteDirectory(source);
+
+      return true;
+    } catch (IOException e) {
+      FileUtils.deleteQuietly(destination);
+      return false;
+    }
+  }
+
+  /**
+   * Detect the path to a home trash directory of the current user.
+   *
+   * <p>The home trash directory should be automatically created for any new user. If the directory
+   * does not exist, it will be created.
+   *
+   * @return the path to the trash directory.
+   */
+  private Path detectTrashDirectory() {
+    var xdgDataHomeOverride = System.getenv(XDG_DATA_HOME);
+    var xdgDataHome =
+        xdgDataHomeOverride == null
+            ? directories.getUserHome().resolve(".local").resolve("share")
+            : Path.of(xdgDataHomeOverride);
+
+    var trashDir = xdgDataHome.resolve(PATH_TRASH);
+
+    try {
+      Files.createDirectories(trashDir.resolve(PATH_FILES));
+    } catch (IOException ignored) {
+    }
+
+    try {
+      Files.createDirectories(trashDir.resolve(PATH_INFO));
+    } catch (IOException ignored) {
+    }
+
+    return trashDir;
+  }
+
+  /**
+   * The trashinfo metadata file.
+   *
+   * @param path the path to this trashinfo file.
+   * @param fileName the file name that should be used to store the trashed file.
+   */
+  private record TrashInfo(Path path, String fileName) {
+
+    private static final int SUFFIX_SIZE = 6;
+    private static final int MAX_ATTEMPTS = Byte.MAX_VALUE;
```

Comment:
Why not just use `127` constant here? Is the maximum number of attempts really tied to the representable range of `Byte`?